File: src/main/java/ru/falseteam/neural2048/NeuralNetworkTrainer.java
package ru.falseteam.neural2048;

import ru.falseteam.neural2048.ga.Fitness;
import ru.falseteam.neural2048.ga.GeneticAlgorithm;
import ru.falseteam.neural2048.ga.IterationListener;
import ru.falseteam.neural2048.ga.MutatorCrossover;
import ru.falseteam.neural2048.gnn.GeneticNeuralNetwork;
import ru.falseteam.neural2048.gnn.crossower.CrossoverAllInOne;
import ru.falseteam.neural2048.gnn.mutations.MutationWeights;
import ru.falseteam.neural2048.logic.GameLogic;
import ru.falseteam.neural2048.nn.NeuralNetwork;
import ru.falseteam.neural2048.players.NeuralNetworkPlayer;

/**
 * Класс для теренеровки нейронной сети.
 *
 * @author Vladialv Sumin
 * @version 1.1
 */
public class NeuralNetworkTrainer implements Fitness<GeneticNeuralNetwork, Integer>,
        IterationListener<GeneticNeuralNetwork, Integer> {
    // НАСТРОЙКИ
    private static final int POPULATION_SIZE = 40;
    private static final int ITERATION = 30;
    private static final int POPULATION_SURVIVE = 5;
    // КОНЕЦ НАСТРОЕК

    private final GameLogic gameLogic;
    private final NeuralNetworkPlayer player;
    private GeneticAlgorithm<GeneticNeuralNetwork, Integer> env;
    private volatile boolean work = false;


    private final Object lock = new Object();
    private final Runnable runner = new Runnable() {
        @Override
        public void run() {
            while (work) {
                env.evolve(1);
            }
        }
    };
    private Thread run;

    public NeuralNetworkTrainer(GameLogic gameLogic, GeneticNeuralNetwork nn) {
        this.gameLogic = gameLogic;
        player = new NeuralNetworkPlayer(null);

        MutatorCrossover<GeneticNeuralNetwork> mutatorCrossover = new MutatorCrossover<>();
        mutatorCrossover.addMutations(new MutationWeights());
        mutatorCrossover.addCrosses(new CrossoverAllInOne());

        env = new GeneticAlgorithm<>(this, mutatorCrossover);

        for (int i = 0; i < POPULATION_SIZE; i++) {
            env.addChromosome(mutatorCrossover.mutate(nn));
        }

        env.addIterationListener(this);
    }

    public void start() {
        synchronized (lock) {
            if (work) return;
            work = true;
            run = new Thread(runner);
            run.start();
        }
    }

    public void stop() {
        synchronized (lock) {
            if (!work) return;
            work = false;
            try {
                run.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Проводит ITERATION игр для каждой хромосомы
     *
     * @return 0 - усредненный счет за ITERATION игр
     */
    @Override
    public Integer calculate(GeneticNeuralNetwork chromosome) {
        int score = 0;
        for (int i = 0; i < ITERATION; i++) {
            player.setNeuralNetwork(chromosome);
            player.playOneGame(gameLogic);
            score += gameLogic.score;
            gameLogic.restart();
        }
        score /= ITERATION;
        return -score;
    }

    private int maxScore = 0;
    private int counter = 0;
    private final int[] scores = new int[10];

    @Override
    public void update(GeneticAlgorithm<GeneticNeuralNetwork, Integer> environment) {
        int score = -environment.getBestFitness();

        System.out.print(environment.getIteration() + "\t");
        System.out.print(score);

        if (maxScore < score) {
            maxScore = score;
            System.out.print("\t(" + maxScore + ")\t NEW RECORD");
        } else
            System.out.print("\t(" + maxScore + ")\t");

        scores[counter++] = score;
        counter %= 10;
        int scoreAvg = 0;
        for (int i = 0; i < 10; i++) {
            scoreAvg += scores[i];
        }
        scoreAvg /= 10;
        System.out.println("\t" + scoreAvg);

        environment.setParentChromosomesSurviveCount(POPULATION_SURVIVE);
    }

    public NeuralNetwork getBest() {
        synchronized (lock) {
            if (work) throw new RuntimeException();
            return env.getBest();
        }
    }
}

File: src/main/java/ru/falseteam/neural2048/NeuralNetworkTrainer.kt
package ru.falseteam.neural2048

import ru.falseteam.neural2048.ga.Fitness
import ru.falseteam.neural2048.ga.GeneticAlgorithm
import ru.falseteam.neural2048.ga.MutatorCrossover
import ru.falseteam.neural2048.gnn.crossower.*
import ru.falseteam.neural2048.gnn.mutations.MutationWeights
import ru.falseteam.neural2048.logic.GameLogic
import ru.falseteam.neural2048.nn.NeuralNetwork
import ru.falseteam.neural2048.nn.NeuralNetworkManager
import ru.falseteam.neural2048.nn.ThresholdFunction
import ru.falseteam.neural2048.players.NeuralNetworkPlayer

/**
 * Класс для теренеровки нейронной сети.
 *
 * @author Vladislav Sumin
 * @version 1.2
 */
class NeuralNetworkTrainer(nn:NeuralNetwork) : Fitness<NeuralNetwork, Int>, GeneticAlgorithm.IterationListener<NeuralNetwork, Int> {

    companion object {
        private val ITERATION = 100
        private val POPULATION_SURVIVE = 5
        private val POPULATION_SIZE = POPULATION_SURVIVE * 4
        private val NEURAL_CONFIG = intArrayOf(16, 32, 16, 4)
        private val thresholdFunctions = arrayOf(ThresholdFunction.SIGMA)//ThresholdFunction.LINEAR,

        /**
         * Созадает тренера по умолчанию
         */
        fun getDefault(): NeuralNetworkTrainer {
            val nn = NeuralNetworkManager.createNeuralNetwork(thresholdFunctions, NEURAL_CONFIG)
            for (i in 0 until 15)
                nn.setNeuronFunction(i, ThresholdFunction.LINEAR, ThresholdFunction.LINEAR.defaultParams)
            for (i in nn.neuronsCount - 4 until nn.neuronsCount - 1)
                nn.setNeuronFunction(i, ThresholdFunction.LINEAR, ThresholdFunction.LINEAR.defaultParams)
            return NeuralNetworkTrainer(nn)
        }
    }

    private val players: Array<NeuralNetworkPlayer> = arrayOf(
            NeuralNetworkPlayer(null, GameLogic(null)),
            NeuralNetworkPlayer(null, GameLogic(null)),
            NeuralNetworkPlayer(null, GameLogic(null)),
            NeuralNetworkPlayer(null, GameLogic(null)))

    private val env: GeneticAlgorithm<NeuralNetwork, Int>
    @Volatile private var work = false


    private val lock = Any()
    private val runner: Runnable
    private var run: Thread? = null

    private var counter = 0
    private val scores = IntArray(10)
    private var time: Long = 0L

    val best: NeuralNetwork
        get() = synchronized(lock) {
            if (work) throw RuntimeException()
            return env.getBest()
        }

    init {
        val mutatorCrossover = MutatorCrossover<NeuralNetwork>()
        mutatorCrossover.addMutations(MutationWeights())
        mutatorCrossover.addCrosses(TwoPointsWeightsCrossover())
        mutatorCrossover.addCrosses(UniformelyDistributedWeightsCrossover())
        mutatorCrossover.addCrosses(TwoPointsNeuronsCrossover())
        mutatorCrossover.addCrosses(UniformelyDistributedNeuronsCrossover())
        env = GeneticAlgorithm(this, mutatorCrossover)
        for (i in 0 until POPULATION_SIZE) {
            env.addChromosome(mutatorCrossover.mutate(nn))
        }
        env.addIterationListener(this)
        env.parentChromosomesSurviveCount = POPULATION_SURVIVE

        runner = Runnable {
            while (work) env.evolve(1)
        }
    }

    fun start() {
        synchronized(lock) {
            if (work) return
            work = true
            time = System.currentTimeMillis()
            run = Thread(runner)
            run!!.start()
        }
    }

    fun stop() {
        synchronized(lock) {
            if (!work) return
            work = false
            try {
                run!!.join()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

        }
    }

    /**
     * Проводит ITERATION игр для каждой хромосомы
     *
     * @return 0 - усредненный счет за ITERATION игр
     */
    override fun calculate(chromosome: NeuralNetwork, threadNumber: Int): Int {
        val player = players[threadNumber]
        var score = 0
        for (i in 0 until ITERATION) {
            player.setNeuralNetwork(chromosome)
            player.playOneGame()
            score += player.gameLogic.score
            player.gameLogic.restart()
        }
        score /= ITERATION
        return -score
    }

    override fun update(environment: GeneticAlgorithm<NeuralNetwork, Int>) {
        val score = -environment.getBestFitness()

        scores[counter++] = score
        counter %= 10
        var scoreAvg = (0..9).sumBy { scores[it] }
        scoreAvg /= 10

        println("${environment.iteration}.    $score (avg $scoreAvg)" +
                " time ${System.currentTimeMillis() - time}ms")
        time = System.currentTimeMillis()
    }
}
